lang/ruby/lib/avro/io/datum_reader_base.rb

```diff
@@ -0,0 +1,95 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+require 'avro/io/json_decoder'
+
+module Avro
+  module IO
+    class DatumReaderBase
+      attr_accessor :writers_schema, :readers_schema
+
+      def initialize(writers_schema=nil, readers_schema=nil)
+        @writers_schema = writers_schema
+        @readers_schema = readers_schema
+      end
+
+      def read(decoder)
+        self.readers_schema = writers_schema unless readers_schema
+      end
+
+      def read_default_value(field_schema, default_value)
+        case field_schema.type_sym
+        when :null;    JsonDecoder.decode_null(default_value)
```

Comment:
nah, not correct.  now the base decoder is using the JsonDecoder.  am I wrong in thinking you can share an entire `reader` object (with no changes), but giving either a `JsonDecoder` or a `BinaryDecoder`?